File: backend/app/schemas/progression.py
"""
Pydantic schemas for the XP / Progression system (Module 12).
"""
from typing import List, Optional
from pydantic import BaseModel


class ProgressionResponse(BaseModel):
    """Full progression snapshot returned to the client."""
    total_xp: int
    level: int
    level_title: str
    xp_in_current_level: int
    xp_for_current_level_span: int
    xp_to_next_level: int
    level_progress_percent: int
    completed_steps_count: int
    completed_projects_count: int
    is_max_level: bool


class XPEventResponse(BaseModel):
    """Returned alongside a project/step update whenever XP may have been awarded."""
    xp_earned: int
    already_awarded: bool
    levelled_up: bool
    new_level: Optional[int] = None
    # Module 13: list of achievement IDs newly unlocked in this event
    newly_unlocked_achievements: List[str] = []
    progression: ProgressionResponse


class ProjectWithXPResponse(BaseModel):
    """
    Envelope used by the step-toggle and project-complete endpoints so the
    frontend can update project state, XP state, and achievements in a single
    round-trip.
    """
    project: dict
    xp_event: XPEventResponse

File: frontend/src/types/progression.ts
// Types for XP / Progression system (Module 12) and Achievements (Module 13)
import type { ProjectItem } from '../services/api/projectApi';

export interface ProgressionData {
  total_xp: number;
  level: number;
  level_title: string;
  xp_in_current_level: number;
  xp_for_current_level_span: number;
  xp_to_next_level: number;
  level_progress_percent: number;
  completed_steps_count: number;
  completed_projects_count: number;
  is_max_level: boolean;
}

export interface XPEvent {
  xp_earned: number;
  already_awarded: boolean;
  levelled_up: boolean;
  new_level: number | null;
  newly_unlocked_achievements: string[];
  progression: ProgressionData;
}

export interface ProjectWithXPResponse {
  project: ProjectItem;
  xp_event: XPEvent;
}

// Achievements

export type AchievementCategory = 'building' | 'exploration' | 'progression';

export interface Achievement {
  id: string;
  name: string;
  description: string;
  icon: string;              // lucide-react icon name
  category: AchievementCategory;
  threshold: number;
  metric: string;
  current_value: number;
  progress_percent: number;
  is_unlocked: boolean;
  unlocked_at: string | null; // ISO-8601 or null
}

export interface AchievementsResponse {
  achievements: Achievement[];
  total_count: number;
  unlocked_count: number;
}